README.md
# ExUnited [![Build Status](https://travis-ci.org/archan937/ex_united.svg?branch=master)](http://travis-ci.org/archan937/ex_united)

Easily spawn Elixir nodes (supervising, Mix configured, easy asserted / refuted) within ExUnit tests

## Introduction

`ExUnited` is a Hex package designed to easily facilitate spawning **supervising**
local Elixir nodes within tests. Unfortunately, I was not able to properly setup
a spawned node for supervisioning with the Erlang `:slave.start_link/1` function.
So I have written `ExUnited` to accomplish that, along with supporting `Mix.Config`
configurations, additional loaded code, and a developer friendly way of writing
assertions and refutations in the context of a spawned node which really improved
the readability of the tests and more.

## Features

  * Spawn nodes for testing purposes
  * Spin up "partially connected" vs "fully connected" nodes
  * Run in "verbose" mode which prints a colorized STDOUT of the nodes
  * Specify extra "code paths" which will be included (`config.exs` included)
  * Support child supervisioning within a spawned node
  * Exclude certain dependencies for spawned nodes
  * Easily(!) assert and refute within the context of spawned nodes

Enjoy the package! I would love to receive a shoutout and/or your feedback ;)

## Installation

To install ExUnited, please add `ex_united` to your list of dependencies in
`mix.exs`:

  ```elixir
  def deps do
    [
      {:ex_united, "~> 0.1.5", only: :test}
    ]
  end
  ```

Replace the default `ExUnit.start()` invocation in the test helper file with
`ExUnited.start()`:

  ```elixir
  # test/test_helper.exs
  ExUnited.start()
  ```

### Explicitly start ExUnit yourself

As of version `0.1.2`, you can also start `ExUnit` yourself explicitly and add
`ExUnited.start(false)` instead:

  ```elixir
  # test/test_helper.exs
  ExUnit.start()
  ExUnited.start(false)
  ```

### ATTENTION: When also using meck-based packages

The following errors can occur when also using packages like
[mock](https://hex.pm/packages/mock) or [MecksUnit](https://hex.pm/packages/mecks_unit)
(which both use the Erlang library [meck](https://github.com/eproxus/meck) to
mock functions) and spawning the nodes with the default environment `test`:

* `(UndefinedFunctionError) function Some.Module.some_function/1 is undefined`
* `(ErlangError) Erlang error: {{:undefined_module, < Some.Module >}`

To tackle this, you should configure any other (Mix) environment to spawn the
nodes with. Configure it like so:

  ```elixir
  # config/test.exs
  import Config

  config :ex_united,
    mix_env: :dev
  ```

You might also want to consider using a bogus environment (e.g. `:void`) to skip
the non-relevant `:dev` dependencies, like `credo` or `dialyxir` probably. That
will save some compile time.

And last but not least, please note that using a different environment within CI
builds will require compiling the project in that particular environment on
beforehand of the tests. Otherwise spawning the nodes will take too much time
and that will cause timeout errors during the tests.

  ```yaml
  # .gitlab-ci.yml
  before_script:
    ...
    - MIX_ENV=void mix deps.get
    - MIX_ENV=void mix run -e 'IO.puts("Done.")'
    - epmd -daemon
  script:
    - mix test
  ```

## Usage

For using `ExUnited`, the two essential functions are:

  1. `ExUnited.spawn/2` - Spawns (`Mix.Config` configured, additional code loaded,
    supervising) nodes
  2. `ExUnited.teardown/0` - Kills the spawned nodes and it also cleans up their
    generated files

### The most simplest setup

Nodes can be specified as a list of atoms, just like in the following example.
Their node names will be `:"bruce@127.0.0.1"` and `:"clark@127.0.0.1"` respectively).

Please do not forget to invoke `ExUnited.teardown/0` at the `on_exit` hook.

  ```elixir
  setup do
    {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

### "Partially versus Fully connected" and/or "Verbose" spawned nodes

As a second argument, you can pass a list of atoms for the options:

  * `:connect` - if `true` a "fully connected" node will be spawned (see
    the `erl -connect_all` flag for more information). Defaults to `false`
  * `:verbose` - if `true` the STDOUT of the spawned node will be printed.
    Defaults to `false`

See `ExUnited.spawn/2` for more information.

  ```elixir
  setup do
    {:ok, spawned} = ExUnited.spawn([:roy], [:connect, :verbose])

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

Which results in the following when running tests:

  ```shell
  PME-Legend ~/S/ex_united:master> mix test test/ex_united/supervised_test.exs:140
  Excluding tags: [:test]
  Including tags: [line: "140"]

  iex(roy@127.0.0.1)> Compiling 1 file (.ex)
  iex(roy@127.0.0.1)> Generated void app
  iex(roy@127.0.0.1)> Interactive Elixir (1.10.1) - press Ctrl+C to exit (type h() ENTER for help)
  iex(roy@127.0.0.1)1>
  .

  Finished in 0.9 seconds
  2 tests, 0 failures, 1 excluded
  ```

### Exclude certain dependencies for all spawned nodes

You can exclude certain (Mix) dependencies for ALL spawned nodes by for instance
adding `exclude: [:inch_ex]` to the options. This can significantly improve
the speed of your tests.

  ```elixir
  setup do
    {:ok, spawned} = ExUnited.spawn([:bruce, :clark], [:verbose, exclude: [:inch_ex]])

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

The following dependencies are excluded by default:

* `:credo`
* `:dialyxir`
* `:ex_doc`
* `:ex_united`
* `:excoveralls`

### Configuring the spawned nodes

Aside from the list of atoms, you can also specify nodes as a keyword list in
case you want to configure them. The following options are available:

* `:code_paths` - a list of directories that will be included
* `:exclude` - a list of dependencies that will be excluded
* `:supervise` - the child spec(s) used for supervisioning

### Including additional code

It would be a best practice to create a directory called `test/nodes` in which
you put a directory containing code for a specific spawned node. Please note that
the file called `config.exs` is supported for `Mix.Config`:

  ```elixir
  setup do
    {:ok, spawned} =
      ExUnited.spawn(
        eric: [
          code_paths: [
            "test/nodes/cantona"
          ]
        ]
      )

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

See [test/ex_united/supervised_test.exs](https://github.com/archan937/ex_united/blob/v0.1.5/test/ex_united/supervised_test.exs#L7)
with its corresponding [test/nodes/ronaldo](https://github.com/archan937/ex_united/tree/v0.1.5/test/nodes/ronaldo)
as an example.

### Exclude certain dependencies for a specific spawned node

Add the `:exclude` list as follows:

  ```elixir
  setup do
    {:ok, spawned} =
      ExUnited.spawn(
        bruce: [
          code_paths: [
            "test/nodes/bruce"
          ],
          exclude: [
            :my_unused_dependency,
            :my_current_project
          ],
          supervise: [MyAwesomeGenServer]
        ],
        clark: [
          code_paths: [
            "test/nodes/clark"
          ],
          supervise: [MyOtherAwesomeGenServer]
        ]
      )

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

### Add supervisioning

Childspecs should be the same argument as if you are adding them to your classic
`<app>/application.ex` file:

  ```elixir
  setup do
    {:ok, spawned} =
      ExUnited.spawn(
        bruce: [
          code_paths: [
            "test/nodes/bruce"
          ],
          supervise: [MyAwesomeGenServer]
        ],
        clark: [
          code_paths: [
            "test/nodes/clark"
          ],
          supervise: [MyOtherAwesomeGenServer]
        ]
      )

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

Pay attention that functions within childspecs should be quoted.

  ```elixir
  setup do
    {:ok, spawned} =
      ExUnited.spawn(
        [
          roy: [
            code_paths: [
              "test/nodes/keane"
            ],
            supervise: [
              {
                Roy,
                talk:
                  quote do
                    fn
                      1 -> "Hi, I am Roy Keane"
                      2 -> "I am keen as mustard"
                      3 -> "I like to be peachy keen"
                    end
                  end
              }
            ]
          ]
        ],
        [:verbose]
      )

    on_exit(fn ->
      ExUnited.teardown()
    end)

    spawned
  end
  ```

### Easily assert and refute within the context of spawned nodes

To seemlessly execute assertions and refutations within spawned nodes, you can
setup your test module by either using `ExUnited.Case` instead of `ExUnit.Case`:

  ```elixir
  defmodule MyNodesTest do
    use ExUnited.Case
  end
  ```

Or by importing the `ExUnited.Case` module:

  ```elixir
  defmodule MyNodesTest do
    use ExUnit.Case
    import ExUnited.Case
  end
  ```

Writing assertions and refutations within the context of a certain spawned is
pretty straight forward with the use of the `ExUnited.Case.as_node/2` function
as if you are writing your class `assert` and/or `refute` statements:

  ```elixir
  defmodule MyNodesTest do
    use ExUnited.Case

    setup do
      {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

      on_exit(fn ->
        ExUnited.teardown()
      end)

      spawned
    end

    test "assertions and refutations within node contexts", spawned do
      bruce = get_in(spawned, [:bruce, :node])

      as_node(bruce) do
        assert :"bruce@127.0.0.1" = Node.self()
        refute :"clark@127.0.0.1" == Node.self()
      end

      as_node(:clark) do
        assert :"clark@127.0.0.1" = Node.self()
        refute :"bruce@127.0.0.1" == Node.self()
      end
    end
  end
  ```

See `ExUnited.Case.as_node/2` for more information.

## Contact me

For support, remarks and requests, please mail me at [pm_engel@icloud.com](mailto:pm_engel@icloud.com).

## License

Copyright (c) 2020 Paul Engel, released under the MIT License

http://github.com/archan937 – http://twitter.com/archan937 – [pm_engel@icloud.com](mailto:pm_engel@icloud.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## v0.1.5

  * Add config for spawn timeout (credits to @thomas9911)

## v0.1.4

  * Add include spawn option (credits to @thomas9911)

## v0.1.3

  * Add necessary configs to run on umbrella projects (credits to @PinheiroRodrigo)

## v0.1.2

  * Add ability to use a different `MIX_ENV` when spawning nodes (solves `meck` related testing problems)
  * Add boolean flag to not start ExUnit

## v0.1.1

  * Add `:exclude` option for individual spawned nodes
  * Exclude current project as dependency (instead, add "lib" to code paths)
  * Do not generate config files
  * Fix redundant prompt when in verbose mode

## v0.1.0

  * Initial commit
    * Spawn nodes for testing purposes
    * Spin up "partially connected" vs "fully connected" nodes
    * Run in "verbose" mode which prints a colorized STDOUT of the nodes
    * Specify extra "code paths" which will be included (`config.exs` included)
    * Support child supervisioning within a spawned node
    * Exclude certain dependencies for spawned nodes
    * Easily(!) assert and refute within the context of spawned nodes


test/test_helper.exs
defmodule TestHelper do
  def take(spawned, key) do
    spawned
    |> Enum.reduce([], fn
      {_name, %{node: _node, port: _port, env: _env} = node}, list ->
        list ++ [Map.get(node, key)]

      _, list ->
        list
    end)
  end
end

ExUnited.start()


test/nodes/keane/roy.ex
defmodule Roy do
  @moduledoc false

  use GenServer

  def start_link(opts) do
    talk = Keyword.get(opts, :talk)
    GenServer.start_link(__MODULE__, %{counter: 0, talk: talk}, name: name())
  end

  defp name, do: Roy.Server

  def init(state), do: {:ok, state}

  def talk do
    GenServer.call(name(), :talk)
  end

  def handle_call(:talk, _from, %{counter: counter, talk: talk} = state) do
    counter = counter + 1
    greet = talk.(counter)

    {:reply, greet, %{state | counter: counter}}
  end
end


mix.exs
defmodule ExUnited.MixProject do
  use Mix.Project

  @version "0.1.5"

  def project do
    [
      app: :ex_united,
      version: @version,
      elixir: ">= 1.6.0",
      elixirc_options: [warnings_as_errors: true],
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      deps: deps(),
      dialyzer: dialyzer(),
      docs: docs(),
      name: "ExUnited",
      preferred_cli_env: [
        credo: :credo,
        dialyzer: :dialyzer,
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test,
        inch: :inch,
        "inch.report": :inch
      ],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp description do
    """
    Easily spawn Elixir nodes (supervising, Mix configured, easy asserted / refuted) within ExUnit tests
    """
  end

  defp package do
    [
      maintainers: ["Paul Engel"],
      licenses: ["MIT"],
      links: %{github: "https://github.com/archan937/ex_united"}
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.3", only: :credo, runtime: false},
      {:dialyxir, "~> 1.0", only: :dialyzer, runtime: false},
      {:ex_doc, "~> 0.21", only: :dev, runtime: false},
      {:excoveralls, "~> 0.12.3", only: :test},
      {:inch_ex, "~> 2.0", only: :inch, runtime: false}
    ]
  end

  defp dialyzer do
    [
      plt_file: {:no_warn, ".ex_united.plt"},
      plt_add_apps: [:mix, :ex_unit, :eex],
      ignore_warnings: ".dialyzer-ignore.exs"
    ]
  end

  defp docs do
    [
      main: "readme",
      extras: [
        "README.md",
        "CHANGELOG.md"
      ],
      source_ref: "v#{@version}",
      source_url: "https://github.com/archan937/ex_united",
      groups_for_modules: [
        Types: [
          ExUnited.Node,
          ExUnited.Spawn.State
        ]
      ]
    ]
  end
end


lib/ex_united/node.ex
defmodule ExUnited.Node do
  @moduledoc """
  A struct containing information about a spawned node for use in ExUnit tests:

    * `:node` - the node name of the spawned node
    * `:port` - the corresponding `Port` reference of the spawned node
    * `:command` - the command used for spawning the node
    * `:env` - the list containing environment variables used spawning the node

  See `ExUnited.Spawn.summon/2` for more information.
  """

  @behaviour Access

  # coveralls-ignore-start

  @doc false
  defdelegate fetch(term, key), to: Map

  @doc false
  defdelegate get(term, key, default), to: Map

  @doc false
  defdelegate get_and_update(term, key, fun), to: Map

  @doc false
  defdelegate pop(map, key), to: Map

  # coveralls-ignore-stop

  @type t :: %__MODULE__{
          node: node,
          port: port,
          command: binary,
          env: [{charlist, charlist}]
        }

  defstruct [:node, :port, :command, env: []]
end


lib/ex_united/spawn/state.ex
defmodule ExUnited.Spawn.State do
  @moduledoc """
  Represents the state of an `ExUnited.Spawn` instance. It stores:

    * `:color_index` - a cycling index used for allocating the output color of
      the next node to be spawned with 'verbose mode' enabled
    * `:nodes` - a map containing information about its spawned nodes
  """

  @behaviour Access

  # coveralls-ignore-start

  @doc false
  defdelegate fetch(term, key), to: Map

  @doc false
  defdelegate get(term, key, default), to: Map

  @doc false
  defdelegate get_and_update(term, key, fun), to: Map

  @doc false
  defdelegate pop(map, key), to: Map

  # coveralls-ignore-stop

  @type t :: %__MODULE__{
          color_index: integer,
          nodes: %{node: spawned_node}
        }

  defstruct color_index: 0, nodes: %{}

  @typedoc """
  A struct that keeps information about a spawned node:

    * `:node` - the node name of the spawned node
    * `:port` - the `Port` reference after having spawned the node
    * `:color` - the ANSI color code portion used when in 'verbose mode'
  """

  @type spawned_node :: %{
          node: node,
          port: port,
          color: binary | nil
        }
end


lib/ex_united.ex
defmodule ExUnited do
  @moduledoc """
  This module facilitates spawning nodes within tests.

  For using `ExUnited`, the two essential functions are:

    1. `ExUnited.spawn/2` - Spawns (`Mix.Config` configured, additional code loaded,
      supervising) nodes
    2. `ExUnited.teardown/0` - Kills the spawned nodes and it also cleans up their
      generated files

  ### The most simplest setup

  Nodes can be specified as a list of atoms, just like in the following example.
  Their node names will be `:"bruce@127.0.0.1"` and `:"clark@127.0.0.1"` respectively).

  Please do not forget to invoke `ExUnited.teardown/0` at the `on_exit` hook.

      setup do
        {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  ### "Partially versus Fully connected" and/or "Verbose" spawned nodes

  As a second argument, you can pass a list of atoms for the options:

    * `:connect` - if `true` a "fully connected" node will be spawned (see
      the `erl -connect_all` flag for more information). Defaults to `false`
    * `:verbose` - if `true` the STDOUT of the spawned node will be printed.
      Defaults to `false`

  See `ExUnited.spawn/2` for more information.

      setup do
        {:ok, spawned} = ExUnited.spawn([:roy], [:connect, :verbose])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Which results in the following when running tests:

      PME-Legend ~/S/ex_united:master> mix test test/ex_united/supervised_test.exs:140
      Excluding tags: [:test]
      Including tags: [line: "140"]

      iex(roy@127.0.0.1)> Compiling 1 file (.ex)
      iex(roy@127.0.0.1)> Generated void app
      iex(roy@127.0.0.1)> Interactive Elixir (1.10.1) - press Ctrl+C to exit (type h() ENTER for help)
      iex(roy@127.0.0.1)1>
      .

      Finished in 0.9 seconds
      2 tests, 0 failures, 1 excluded

  ### Exclude certain dependencies within spawned nodes

  You can exclude certain (Mix) dependencies from your spawned nodes by for instance
  adding `exclude: [:inch_ex]` to the options. This can significantly improve
  the speed of your tests.

      setup do
        {:ok, spawned} = ExUnited.spawn([:bruce, :clark], [:verbose, exclude: [:inch_ex]])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  The following dependencies are excluded by default:

  * `:credo`
  * `:dialyxir`
  * `:ex_doc`
  * `:ex_united`
  * `:excoveralls`

  ### Configuring the spawned nodes

  Aside from the list of atoms, you can also specify nodes as a keyword list in
  case you want to configure them. The following options are available:

  * `:code_paths` - a list of directories that will be included
  * `:exclude` - a list of dependencies that will be excluded
  * `:supervise` - the child spec(s) used for supervisioning

  ### Including additional code

  It would be a best practice to create a directory called `test/nodes` in which
  you put a directory containing code for a specific spawned node. Please note that
  the file called `config.exs` is supported for `Mix.Config`:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            eric: [
              code_paths: [
                "test/nodes/cantona"
              ]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  See [test/ex_united/supervised_test.exs](https://github.com/archan937/ex_united/blob/v0.1.5/test/ex_united/supervised_test.exs#L7)
  with its corresponding [test/nodes/ronaldo](https://github.com/archan937/ex_united/tree/v0.1.5/test/nodes/ronaldo)
  as an example.

  ### Exclude certain dependencies for a specific spawned node

  Add the `:exclude` list as follows:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            bruce: [
              code_paths: [
                "test/nodes/bruce"
              ],
              exclude: [
                :my_unused_dependency
              ],
              supervise: [MyAwesomeGenServer]
            ],
            clark: [
              code_paths: [
                "test/nodes/clark"
              ],
              supervise: [MyOtherAwesomeGenServer]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  ### Add supervisioning

  Childspecs should be the same argument as if you are adding them to your classic
  `<app>/application.ex` file:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            bruce: [
              code_paths: [
                "test/nodes/bruce"
              ],
              supervise: [MyAwesomeGenServer]
            ],
            clark: [
              code_paths: [
                "test/nodes/clark"
              ],
              supervise: [MyOtherAwesomeGenServer]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Pay attention that functions within childspecs should be quoted.

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            [
              roy: [
                code_paths: [
                  "test/nodes/keane"
                ],
                supervise: [
                  {
                    Roy,
                    talk:
                      quote do
                        fn
                          1 -> "Hi, I am Roy Keane"
                          2 -> "I am keen as mustard"
                          3 -> "I like to be peachy keen"
                        end
                      end
                  }
                ]
              ]
            ],
            [:verbose]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  ### Easily assert and refute within the context of spawned nodes

  To seemlessly execute assertions and refutations within spawned nodes, you can
  setup your test module by either using `ExUnited.Case` instead of `ExUnit.Case`:

      defmodule MyNodesTest do
        use ExUnited.Case
      end

  Or by importing the `ExUnited.Case` module:

      defmodule MyNodesTest do
        use ExUnit.Case
        import ExUnited.Case
      end

  Writing assertions and refutations within the context of a certain spawned is
  pretty straight forward with the use of the `ExUnited.Case.as_node/2` function
  as if you are writing your class `assert` and/or `refute` statements:

      defmodule MyNodesTest do
        use ExUnited.Case

        setup do
          {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

          on_exit(fn ->
            ExUnited.teardown()
          end)

          spawned
        end

        test "assertions and refutations within node contexts", spawned do
          bruce = get_in(spawned, [:bruce, :node])

          as_node(bruce) do
            assert :"bruce@127.0.0.1" = Node.self()
            refute :"clark@127.0.0.1" == Node.self()
          end

          as_node(:clark) do
            assert :"clark@127.0.0.1" = Node.self()
            refute :"bruce@127.0.0.1" == Node.self()
          end
        end
      end

  See `ExUnited.Case.as_node/2` for more information.

  In some cases the summon of a node takes longer than five seconds. This can be changed in your config:
  ```elixir
  # in config/test.exs
  config :ex_united,
    summon_timeout: 15000
  ```
  """

  @emptyconfig Path.expand(
                 "../ex_united/config.exs",
                 __ENV__.file
               )

  @excluded_dependencies [
    :credo,
    :dialyxir,
    :ex_doc,
    :ex_united,
    :excoveralls
  ]

  @included_dependencies []

  alias ExUnited.Node, as: ExNode
  alias ExUnited.Spawn, as: Simmons

  @doc """
  Starts both the `ExUnit` and `ExUnited.Spawn` gen servers. This should replace
  the default `ExUnit.start()` invocation in the test helper file.

      # test/test_helper.exs
      ExUnited.start()

  As of version `0.1.2`, you can also start `ExUnit` yourself explicitly and add
  `ExUnited.start(false)` instead:

      # test/test_helper.exs
      ExUnit.start()
      ExUnited.start(false)

  """
  @spec start(boolean) :: {:ok, pid}
  def start(start_ex_unit \\ true) do
    if start_ex_unit do
      ExUnit.start()
    end

    Simmons.start_link()
  end

  @doc """
  Spawns nodes for testing purposes. Supervised applications are supported.

  Nodes can be either specified as a list of atoms (such as `[:bruce, :clark]` for
  instance; the node names will be `:"bruce@127.0.0.1"` and `:"clark@127.0.0.1"`
  respectively) or as a keyword list (in case of configuring the spawned node).

  The following options are available to configure nodes:

    * `:code_paths` - a list of directories that will be included (please note
      that the file called `config.exs` is supported for `Mix.Config`)
    * `:exclude` - a list of dependencies that will be excluded
    * `:include` - a list of dependencies that will be included, in the format that is used in the mix file
    * `:supervise` - the child spec(s) used for supervisioning

  Aside from options for configuring individual nodes, as a second argument, you
  can pass a list of atoms for the following:

    * `:connect` - if `true` a "fully connected" node will be spawned (see
      the `erl -connect_all` flag for more information). Defaults to `false`
    * `:verbose` - if `true` the STDOUT of the spawned node will be printed.
      Defaults to `false`

  And last but not least, you can exclude certain (Mix) dependencies from your
  spawned nodes by adding `exclude: [:inch_ex]` to the options. This can
  significantly improve the speed of your tests.

  The following dependencies are excluded by default:

  * `:credo`
  * `:dialyxir`
  * `:ex_doc`
  * `:ex_united`
  * `:excoveralls`

  ## Examples

  The most simplest setup:

      setup do
        {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Spawn "fully connected" nodes and print all their STDOUT in the console:

      setup do
        {:ok, spawned} = ExUnited.spawn([:bruce, :clark], [:connect, :verbose])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Exclude certain dependencies for all nodes:

      setup do
        {:ok, spawned} = ExUnited.spawn([:bruce, :clark], [:verbose, exclude: [:inch_ex]])

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  A configured nodes setup:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            bruce: [
              code_paths: [
                "test/nodes/bruce"
              ],
              supervise: [MyAwesomeGenServer]
            ],
            clark: [
              code_paths: [
                "test/nodes/clark"
              ],
              supervise: [MyOtherAwesomeGenServer]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Exclude certain dependencies for a specific spawned node:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            bruce: [
              code_paths: [
                "test/nodes/bruce"
              ],
              exclude: [
                :my_unused_dependency
              ],
              supervise: [MyAwesomeGenServer]
            ],
            clark: [
              code_paths: [
                "test/nodes/clark"
              ],
              supervise: [MyOtherAwesomeGenServer]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Include certain dependencies for a specific spawned node:

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            bruce: [
              code_paths: [
                "test/nodes/bruce"
              ],
              include: [
                {:my_lib, ">= 0.0.0"},
                {:another_lib, "~> 1.0"}
              ],
              supervise: [MyAwesomeGenServer]
            ],
            clark: [
              code_paths: [
                "test/nodes/clark"
              ],
              supervise: [MyOtherAwesomeGenServer]
            ]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end

  Also note that functions within childspecs should be quoted.

      setup do
        {:ok, spawned} =
          ExUnited.spawn(
            [
              roy: [
                code_paths: [
                  "test/nodes/keane"
                ],
                supervise: [
                  {
                    Roy,
                    talk:
                      quote do
                        fn
                          1 -> "Hi, I am Roy Keane"
                          2 -> "I am keen as mustard"
                          3 -> "I like to be peachy keen"
                        end
                      end
                  }
                ]
              ]
            ],
            [:verbose]
          )

        on_exit(fn ->
          ExUnited.teardown()
        end)

        spawned
      end
  """
  @spec spawn([node] | [{node, keyword}], [atom]) :: {:ok, [ExNode.t()]}
  def spawn(nodes, opts \\ []) do
    spawned =
      nodes
      |> Enum.map(fn node ->
        {name, spec} =
          case node do
            {_name, _spec} -> node
            name -> {name, []}
          end

        generate_mix_exs(name, opts, spec)
        {name, spawn_node(name, opts)}
      end)
      |> Enum.into(%{})

    {:ok, spawned}
  end

  @doc """
  Should be invoked at the end of a test which spawned nodes. This kills the
  nodes and it also cleans up their generated files located in
  `/tmp/[NODENAME]-{config,mix}.exs`.

  ## Example

      defmodule MyNodesTest do
        use ExUnit.Case

        setup do
          {:ok, spawned} = ExUnited.spawn([:bruce, :clark])

          on_exit(fn ->
            ExUnited.teardown()
          end)

          spawned
        end

        test "does awesome stuff with spawned nodes", spawned do
          # a lot of awesome assertions and refutations
        end
      end
  """
  @spec teardown() :: :ok
  def teardown do
    Simmons.kill_all()

    "/tmp/*-{config,mix}.exs"
    |> Path.wildcard()
    |> Enum.each(&File.rm/1)

    :ok
  end

  @spec generate_mix_exs(atom, [atom | keyword], keyword) :: :ok
  defp generate_mix_exs(name, opts, spec) do
    name
    |> mix_exs_path()
    |> File.write(mix_exs(opts, spec))

    :ok
  end

  @spec mix_exs_path(atom) :: binary
  defp mix_exs_path(name), do: "/tmp/#{name}-mix.exs"

  @spec mix_exs([atom | keyword], keyword) :: term
  def mix_exs(opts, spec) do
    config = Mix.Project.config()

    project =
      config
      |> Keyword.take([:version, :elixir, :build_path, :deps_path, :lockfile])
      |> Keyword.put(:app, :void)
      |> Keyword.put(:config_path, @emptyconfig)
      |> Keyword.put(:elixirc_paths, elixirc_paths(opts, spec))
      |> Keyword.put(:deps, deps(config, opts, spec))

    "../ex_united/mix.exs.eex"
    |> Path.expand(__ENV__.file)
    |> EEx.eval_file(
      project: project,
      all_env: read_config(opts, spec),
      supervised: supervised(spec)
    )
  end

  @spec elixirc_paths(keyword, keyword) :: list
  defp elixirc_paths(opts, spec) do
    app = Keyword.get(Mix.Project.config(), :app)

    exclude =
      @excluded_dependencies ++
        List.wrap(Keyword.get(opts, :exclude)) ++
        List.wrap(Keyword.get(spec, :exclude))

    code_paths =
      if Enum.member?(exclude, app) do
        []
      else
        ["lib"]
      end

    code_paths ++ Keyword.get(spec, :code_paths, [])
  end

  @spec deps(keyword, [atom | keyword], keyword) :: list
  defp deps(config, opts, spec) do
    exclude =
      @excluded_dependencies ++
        List.wrap(Keyword.get(opts, :exclude)) ++
        List.wrap(Keyword.get(spec, :exclude))

    include =
      @included_dependencies ++
        List.wrap(Keyword.get(opts, :include)) ++
        List.wrap(Keyword.get(spec, :include))

    config
    |> Keyword.get(:deps)
    |> Enum.reject(fn dep ->
      Enum.member?(exclude, elem(dep, 0))
    end)
    |> Enum.concat(include)
  end

  @spec read_config(keyword, keyword) :: keyword
  defp read_config(opts, spec) do
    config =
      opts
      |> elixirc_paths(spec)
      |> Enum.map(fn dir ->
        Path.wildcard("#{dir}/config.exs")
      end)
      |> List.flatten()
      |> case do
        [config] -> config
        [] -> @emptyconfig
      end

    if function_exported?(Config.Reader, :read!, 1) do
      Config.Reader
    else
      Mix.Config
    end
    |> apply(:read!, [config])
  end

  @spec supervised(keyword) :: binary
  defp supervised(spec) do
    case Keyword.get(spec, :supervise) do
      nil ->
        "[]"

      children ->
        Macro.to_string(children)
    end
  end

  @spec spawn_node(atom, [atom]) :: ExNode.t()
  defp spawn_node(name, opts) do
    mix_env = Application.get_env(:ex_united, :mix_env, Mix.env())

    {node, port, command, env} =
      Simmons.summon(
        name,
        env: [MIX_ENV: mix_env, MIX_EXS: mix_exs_path(name)],
        connect: Enum.member?(opts, :connect),
        verbose: Enum.member?(opts, :verbose)
      )

    %ExNode{node: node, port: port, command: command, env: env}
  end
end
